ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt

```diff
@@ -0,0 +1,221 @@
+import java.util.concurrent.atomic.*
+
+
+/**
+ * Enables Micrometer support when installed. Exposes the following metrics:
+ * <ul>
+ *     <li><code>ktor.http.server.requests.active</code>: Gauge - The amount of active ktor requests</li>
+ *     <li><code>ktor.http.server.requests</code>: Timer - Timer for all requests. By default no percentiles or
+ *       histogram is exposed. Use the [Configuration.distributionStatisticConfig] to enable these.
+ *       Tags by default (use [Configuration.tags] to configure the tags or add custom tags):
+ *       <ul>
+ *           <li><code>local</code>: The host and port of the request uri (e.g. 'www.ktor.io')</li>
+ *           <li><code>method</code>: The http method (e.g. 'GET')</li>
+ *           <li><code>route</code>: The use ktor route used for this request. (e.g. '/some/path/{someParameter}')
+ *           <li><code>status</code>: The http status code or the Exception name thrown during the request (e.g. '200', 'java.lang.IllegalArgumentException')
+ *        <ul>
+ *     <li>
+ *  <ul>
+ */
+class MicrometerMetrics(
+    private val registry: MeterRegistry,
+    timerDistributionConfig: DistributionStatisticConfig,
+    private val tags: MicrometerMetrics.TagBuilder.() -> Unit
+) {
+
+    private val active = registry.gauge(activeGaugeName, AtomicInteger(0))
+
+    init {
+        enableTimerDistributionConfig(timerDistributionConfig)
+    }
+
+    private fun enableTimerDistributionConfig(timerDistributionConfig: DistributionStatisticConfig) {
+        registry.config().meterFilter(object : MeterFilter {
+            override fun configure(id: Meter.Id, config: DistributionStatisticConfig): DistributionStatisticConfig =
+                if (id.name == requestTimerName)
+                    timerDistributionConfig.merge(config)
+                else
+                    config
+        })
+    }
+
+    /**
+     * Configures this Feature
+     * @property registry The meter registry where the meters are registered. Mandatory
+     * @property meterBinders The binders that are automatically bound to the registry. Default: [ClassLoaderMetrics],
+     * [JvmMemoryMetrics], [ProcessorMetrics], [JvmGcMetrics], [ProcessorMetrics], [JvmThreadMetrics], [FileDescriptorMetrics]
+     * @property distributionStatisticConfig configures the histogram and/or percentiles for all request timers. By
+     * default 50%, 90% , 95% and 99% percentiles are configured
+     * @property tags is called to generate the tags for this call. By default, [TagBuilder.defaultTags] are generated
+     */
+    class Configuration {
+
+        lateinit var registry: MeterRegistry
+
+        var meterBinders: List<MeterBinder> = listOf(
+            ClassLoaderMetrics(),
+            JvmMemoryMetrics(),
+            JvmGcMetrics(),
+            ProcessorMetrics(),
+            JvmThreadMetrics(),
+            FileDescriptorMetrics()
+        )
+
+        var distributionStatisticConfig: DistributionStatisticConfig =
```

Comment:
Since Micrometer provides a `MeterFilter` facility to alter the distribution statistic config on any meter, I don't know that it's necessary to include one here as well. One way the framework can help is by binding some set of properties to a meter filter, like Spring Boot does [here](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#_per_meter_properties).

Reply:
What I dont like about the spring approach is that you 
a) have to switch the language (from Java/Kotlin to properties), where you dont have the possibility to let it validate by the browser.
b) from there you have a static config for all DistributionStatistics (no matter where it is used in your app. Imagine you have a timer for fastrunning web requests and one for a long running batch job. Both would need the same value for max expected value.

Reply:
BTW, it's not possible to configure it per route. Does micrometer backends provide ability to see stats per rote/mask?